R/mutate.R

```diff
@@ -0,0 +1,162 @@
+#' rec <- 
+#'   recipe( ~ ., data = iris) %>%
+#'   step_mutate(
+#'     dbl_width = Sepal.Width * 2,
+#'     half_length = Sepal.Length / 2
+#'   )
+#' 
+#' prepped <- prep(rec, training = iris %>% slice(1:75), retain = TRUE)
+#' 
+#' library(dplyr)
+#' 
+#' dplyr_train <- 
+#'   iris %>%
+#'   as_tibble() %>%
+#'   slice(1:75) %>%
+#'   mutate(
+#'     dbl_width = Sepal.Width * 2,
+#'     half_length = Sepal.Length / 2
+#'   )
+#' 
+#' rec_train <- juice(prepped)
+#' all.equal(dplyr_train, rec_train)
+#' 
+#' dplyr_test <- 
+#'   iris %>%
+#'   as_tibble() %>%
+#'   slice(76:150) %>%
+#'   mutate(
+#'     dbl_width = Sepal.Width * 2,
+#'     half_length = Sepal.Length / 2
+#'   )
+#' rec_test <- bake(prepped, iris %>% slice(76:150))
+#' all.equal(dplyr_test, rec_test)
+#' 
+#' # Embedding objects:
+#' const <- 1.414
+#' 
+#' qq_rec <- 
+#'   recipe( ~ ., data = iris) %>%
+#'   step_mutate(
+#'     bad_approach = Sepal.Width * const,
+#'     best_approach = Sepal.Width * !!const
+#'   ) %>%
+#'   prep(training = iris, retain = TRUE)
+#' 
+#' juice(qq_rec, contains("appro")) %>% slice(1:4)
+#' 
+#' # The difference:
+#' tidy(qq_rec, number = 1)  
+
+step_mutate <- function(
+  recipe, ..., 
+  role = "predictor", 
+  trained = FALSE, 
+  input = NULL,
+  skip = FALSE,
+  id = rand_id("mutate")
+) {
+  
+  inputs <- enquos(...)
+  if (is_empty(inputs)) 
+    stop("Please supply at least one name-value pair.", call. = FALSE)
+  if (any(names(inputs) == ""))
+    stop("Please assign new variables to names.", call. = FALSE)
+  
+  add_step(
+    recipe, 
+    step_mutate_new(
+      terms = terms, 
+      trained = trained,
+      role = role, 
+      input = inputs,
+      skip = skip,
+      id = id
+    )
+  )
+}
+
+step_mutate_new <- 
+  function(terms, role, trained, input, skip, id) {
+    step(
+      subclass = "mutate", 
+      terms = terms,
+      role = role,
+      trained = trained,
+      input = input,
+      skip = skip,
+      id = id
+    )
+  }
+
+#' @export
+prep.step_mutate <- function(x, training, info = NULL, ...) {
+  step_mutate_new(
+    terms = x$terms, 
+    trained = TRUE,
+    role = x$role, 
+    input = x$input,
+    skip = x$skip,
+    id = x$id
+  )
+}
+
+#' @importFrom dplyr mutate
+#' @export
+bake.step_mutate <- function(object, newdata, ...) {
+  dplyr::mutate(newdata, !!!object$input)
+}
+
+
+print.step_mutate <-
+  function(x, width = max(20, options()$width - 35), ...) {
+    cat("Variable mutation for ", 
+        paste0(names(x$input), collapse = ", "))
+    if (x$trained) {
+      cat(" [trained]\n") 
+    } else {
+      cat("\n")
+    }
+    invisible(x)
+  }
+
+#' @importFrom rlang quo_get_expr expr_text
+#' @importFrom purrr map map_chr
+#' @importFrom dplyr tibble
+#' @rdname step_mutate
+#' @param x A `step_mutate` object
+#' @export
+tidy.step_mutate <- function(x, ...) {
+  var_expr <- map(x$input, quo_get_expr)
+  var_expr <- map_chr(var_expr, expr_text, width = options()$width, nlines = 1)
```

Comment:
Should be `quo_text(x)` or `expr_text(quo_squash(x))`. Unwrapping the quosure is not sufficient, there might be nested quosures in the expression. The rlang deparsing API is not great, in particular the duplication between `expr_` and `quo_` functions is a bit silly. We'll review it for rlang 0.4.0.

Reply:
done